### PR TITLE
Refactor Enketo service code to allow for reuse by test harness

### PR DESCRIPTION
**Pull request by jkuester**
Refactors code for Enketo forms to allow for reuse by https://github.com/medic/cht-conf-test-harness.  The goal of these changes is to minimize the amount of duplicate code required in the test harness for working with Enketo forms. This should make it easier to maintain the test harness going forwards (simplifying the process of uplifting to new versions of Enketo/cht-core).

@kennsippell has been working on integrating the test harness with this refactored code in https://github.com/medic/cht-conf-test-harness/tree/enketo-uplift.

Changelist:

- api/
  - The `generate-xform` service contained logic both for generating the xForm html files and for updating the forms stored on the client when one changed. This update logic was discrete from the generation logic and had some dependencies that caused issues for the test harness code when trying to consume the generation logic. So, I have refactored the update logic into a new `update-xform` service. 
- webapp/
  - The logic in the `enketo` Angular service is the main bit that we want to be able to reuse in the test harness. To do this, I had to decouple it from all the Angular-specific code and services.  I created the `enketo-form-manager.js` which contains the refactored logic and can be pulled in by both the test harness and the `enketo` service.
  - The `enketo-prepopulation-data` service was only used by the `enketo` service, so I have refactored it into the `enketo-data-prepopulator.js` that is consumed directly by `enketo-form-manager`.
  - The `xpath-element-path` provider was only used by the `enketo` service, so I have refactored it into `xpath-element-path.js` that is consumed directly by `enketo-form-manager`.
  - The logic in the `contact-save` service is also needed by the test harness (to save their contact forms). So, I have refactored this logic into `contact-saver.js`.
    - The `EnketoFormManager` exposes the `contact-saver` functionality via the `saveContactForm` method to give a single point of reference for rendering and saving all Enketo forms.
    - I have updated the `enketo` service to have a `saveContactForm` method that can be used by the `contacts-edit` component to save the contact form (instead of using the (now removed) `contact-save` service).
  - The logic in the `enketo-translation` service was only being used by code that is now centralized in the `enketo-form-manager`. I have refactored that logic into a `enketo-data-translator.js` which is referenced directly by enketo logic.

#7462

# Code review checklist

- [x] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [x] Tested: Unit and/or e2e where appropriate
- [x] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -135,7 +135,7 @@
   "bundlesize": [
     {
       "path": "./build/ddocs/medic/_attachments/main.js",
-      "maxSize": "950 kB"
```
Comment (jkuester) [inline package.json:138]: >  PASS  ./build/ddocs/medic/_attachments/main.js: 950.2KB < maxSize 955KB (gzip) 

Unfortunately, we were right below the `950` line before these changes and have just bumped it over...

**Review by garethbowen — COMMENTED** [review]
Discussed in person but adding here to close the loop...

This will definitely reduce complexity for the test harness and make it easier to keep up to date, it does add complexity to cht-core. Furthermore this could deliver more benefit to the harness is it used stronger APIs and was published as a versioned package. Alternatively the harness could be changed to be more like a easy startup script for cht-core so it would test the actual code rather than a subset thereof.

We'd like to spend more time to analyse these options to deliver value for the harness without addition undue complexity to cht-core, particularly as 4.0.0 is a complex enough change already.

With the current focus on getting 4.0.0 stable it was decided to park this PR for now. We can always merge it later with or without improvements. In the meantime the harness can use this branch, and Product will ensure this is rebased periodically to keep it consistent with main.

**Comment by kennsippell** [comment]
With the completion of this PR basically 100% of the code used by the harness would be in `shared-libs`. As far as I understood the plan, this Pr has the potential to be the last step toward a significant milestone of harmonization between harness and core.

> Furthermore this could deliver more benefit to the harness .... as a versioned package.

This value here feels marginal.

> Alternatively the harness could be changed to be more like a easy startup script for cht-core so it would test the actual code rather than a subset thereof.

The app services team continues as the steward of the harness and we welcome your contributions and suggestions including architecture changes.  This is a pretty big change and I have some concerns, specifically:

1. Setup time for tests
2. Test performance
3. Test reliability
4. Process for debugging test logic
5. Cost and staffing of this initiative (vs the benefits?)
6. Backward compatibility and migration of existing test suites

Please ensure that your architecture proposal brings clear data and evidence for these scenarios. Based on the ongoing activities which are bringing clear value to users, I imagine this process is at least a year out.

In the meantime, I'd recommend that we complete the current milestone of harmonization through cht-core shared-libs. If Enketo code is in share-libs (as it was previously in this PR), that is the **last** significant body of shared code and we will be in a significantly improvement maintenance state between harness and core.

**Comment by jkuester** [comment]
This PR will not be merged, but this `enketo-service-refactor` branch will serve as a temporary integration for the `4.0.0`-compatible version of cht-conf-test-harness while we work on a more maintainable solution (hopefully coming in `4.1.0`).

**Pull request closed without merging**